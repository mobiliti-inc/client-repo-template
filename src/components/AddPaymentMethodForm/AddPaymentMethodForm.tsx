import React, { PureComponent } from 'react';
import { card } from 'creditcards';
import styles from './AddPaymentMethodForm.scss';
import TextInput, { INPUT_TYPES } from '../TextInput/TextInput';
import cardExpiryDateFormatter from '../../utils/cardExpiryDateFormatter';
import Button, { BUTTON_TYPES } from '../Button/Button';
import stripeService from '../../services/stripeService';
type AddPaymentMethodFormProps = {
	onFormValidCheck?: (...args: any[]) => any,
	addPaymentCard: (...args: any[]) => any,
	cancelForm: (...args: any[]) => any,
	displayCancelButton?: boolean,
	isPaymentCardAdded?: boolean
};
type AddPaymentMethodFormState = {
	cardNumber: string,
	cardExpirationDate: string,
	cvv: string,
	zip: string,
	formIsValid: boolean,
	error: string,
	showLoader: boolean
};
type AddPaymentMethodFormState = {
	cardNumber: string,
	cardExpirationDate: string,
	cvv: string,
	zip: string,
	formIsValid: boolean,
	error: string,
	showLoader: boolean
};
class AddPaymentMethodForm extends PureComponent<{}, AddPaymentMethodFormState> {
	state = {
		cardNumber: '',
		cardExpirationDate: '',
		cvv: '',
		zip: '',
		formIsValid: false,
		error: '',
		showLoader: false
	};
	componentDidUpdate = prevProps => {
		const { isPaymentCardAdded } = this.props;
		this.isFormValid();
		if (prevProps.isPaymentCardAdded !== isPaymentCardAdded) {
			this.setState({
				cardNumber: '',
				cardExpirationDate: '',
				cvv: '',
				zip: ''
			});
		}
	};
	fields = {
		cardNumber: { maxLength: 16, formatter: value => card.format(value) },
		zip: { maxLength: 6, formatter: value => value },
		cvv: { maxLength: 3, formatter: value => value },
		cardExpirationDate: { maxLength: 5, formatter: value => cardExpiryDateFormatter(value) }
	};
	handleInputChange = event => {
		const { name, value } = event.target;
		if (this.fields[name].maxLength < value.replace(/ /g, '').length) {
			return null;
		}
		if (value.length && name !== 'cardExpirationDate' && !/^\d+$/.test(value.replace(/ /g, ''))) {
			return null;
		}
		this.setState({
			[name]: this.fields[name].formatter(value.replace(/ /g, '')),
			error: ''
		});
	};
	formatCardDetail = (key, value) => {
		this.setState({
			[key]: this.fields[key].formatter(value.replace(/ /g, ''))
		});
	};
	isFormValid = () => {
		const { zip, cvv, cardNumber, cardExpirationDate } = this.state;
		const formIsValid = !!(zip.length && cvv.length && cardNumber.length && cardExpirationDate.length === 5);
		this.setState({
			formIsValid
		});
		this.props.onFormValidCheck({ cardNumber, formIsValid });
		return formIsValid;
	};
	handleSavePaymentCard = () => {
		this.setState({
			showLoader: true
		});
		const { zip, cvv, cardNumber, cardExpirationDate } = this.state;
		stripeService.createToken(
			{
				number: Number(cardNumber.replace(/ /g, '')),
				cvc: Number(cvv.replace(/ /g, '')),
				exp_month: Number(cardExpirationDate.slice(0, 2)),
				exp_year: Number(cardExpirationDate.slice(-2)),
				address_zip: Number(zip)
			},
			(status, data) => {
				if (status === 200) {
					this.setState({
						showLoader: false
					});
					return this.props.addPaymentCard(data.id);
				}
				this.setState({
					error: data.error.message,
					showLoader: false
				});
			}
		);
	};
	render() {
		const { zip, cvv, cardNumber, cardExpirationDate, formIsValid, showLoader, error } = this.state;
		const { displayCancelButton } = this.props;
		const buttonDisabledClass = !formIsValid ? 'btn-disabled' : '';
		return (
			<div styleName="form-container">
				<TextInput
					placeholder="Card Number"
					label="Card Number"
					onChange={this.handleInputChange}
					type={INPUT_TYPES.TEXT}
					value={cardNumber}
					name="cardNumber"
					id="cardNumber"
					isValid
					onBlur={() => this.formatCardDetail('cardNumber', cardNumber)}
				/>
				<div styleName="column-2">
					<TextInput
						placeholder="Expiration"
						label="Expiration"
						onChange={this.handleInputChange}
						type={INPUT_TYPES.TEXT}
						value={cardExpirationDate}
						name="cardExpirationDate"
						id="cardExpirationDate"
						isValid
						onBlur={() => this.formatCardDetail('cardExpirationDate', cardExpirationDate)}
					/>

					<TextInput
						placeholder="Zip"
						label="Zip"
						onChange={this.handleInputChange}
						type={INPUT_TYPES.TEXT}
						value={zip}
						name="zip"
						id="zip"
						isValid
					/>
				</div>
				<div styleName="column-2">
					<TextInput
						placeholder="CVV"
						label="CVV"
						onChange={this.handleInputChange}
						type={INPUT_TYPES.TEXT}
						value={cvv}
						name="cvv"
						id="cvv"
						isValid
					/>
				</div>
				<div>{error.length > 0 && <label styleName="input-validation-message-label">{error}</label>}</div>
				<div styleName="column-2 btn-container">
					<div styleName={`${buttonDisabledClass}`}>
						<Button
							type={BUTTON_TYPES.STANDARD}
							showLoader={showLoader}
							onClick={() => this.handleSavePaymentCard()}
							disabled={!formIsValid}
							bordered
						>
							Save
						</Button>
					</div>
					{displayCancelButton && (
						<Button type={BUTTON_TYPES.STANDARD} disabled={false} onClick={this.props.cancelForm} bordered>
							Cancel
						</Button>
					)}
				</div>
			</div>
		);
	}
}
AddPaymentMethodForm.defaultProps = {
	onFormValidCheck: value => value,
	displayCancelButton: false,
	isPaymentCardAdded: false
};
export default CSSModules(AddPaymentMethodForm, styles, { allowMultiple: true });
