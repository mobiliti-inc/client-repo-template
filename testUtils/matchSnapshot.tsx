const { shallowToJson } = require('enzyme-to-json');

export const matchSnapshot = (subject: any) => {
	expect(shallowToJson(subject)).toMatchSnapshot();
};
