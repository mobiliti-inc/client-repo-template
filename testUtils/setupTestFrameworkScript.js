const React = require("react");
// const CSSModules = require("react-css-modules");
const Enzyme = require("enzyme");
const Adapter = require("enzyme-adapter-react-16");

const { shallow, render, mount } = Enzyme;

Enzyme.configure({ adapter: new Adapter() });
jest.setTimeout(30000);

global.React = React;
// global.CSSModules = CSSModules;
global.shallow = shallow;
global.render = render;
global.mount = mount;
