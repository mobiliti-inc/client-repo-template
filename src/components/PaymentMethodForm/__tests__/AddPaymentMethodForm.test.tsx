import React from 'react';
import { shallow } from 'enzyme';

import { matchSnapshot } from '../../../../testUtils/matchSnapshot';

import AddPaymentMethodForm from '../PaymentMethodForm';

describe('AddPaymentMethodForm', () => {
	const initialProps = {
		onFormValidCheck: jest.fn(),
		addPaymentCard: jest.fn(),
		cancelForm: jest.fn(),
		isPaymentCardAdded: false
	};

	it('renders', () => {
		const tree = shallow(<AddPaymentMethodForm {...initialProps} />);
		matchSnapshot(tree);
	});

	it('renders without onFormValidCheck function', () => {
		const tree = shallow(<AddPaymentMethodForm
			cancelForm={jest.fn()}
			addPaymentCard={jest.fn()}
			displayCancelButton
			isPaymentCardAdded
		/>);
		matchSnapshot(tree);
	});

	// it('sets state properly when handleInputChange is called with value not set to empty', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.cardNumber).toEqual('');
	// 	tree.instance().handleInputChange({ target: { value: '234567890876', name: 'cardNumber' } });
	// 	expect(tree.instance().state.cardNumber).toEqual('2345 6789 0876');
	// });

	// it('does not change state when handleInputChange is called with value set to non-digit value', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.cardNumber).toEqual('');
	// 	tree.instance().handleInputChange({ target: { value: 'aaaa', name: 'cardNumber' } });
	// 	expect(tree.instance().state.cardNumber).toEqual('');
	// });

	// it('does not change state when handleInputChange is called with value greater than required max length', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.cardNumber).toEqual('');
	// 	tree.instance().handleInputChange({ target: { value: '234567890876456786567865', name: 'cardNumber' } });
	// 	expect(tree.instance().state.cardNumber).toEqual('');
	// });

	// it('formats cardNumber state properly when formatCardDetail', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.cardNumber).toEqual('');
	// 	tree.instance().formatCardDetail('cardNumber', '234567890876');
	// 	expect(tree.instance().state.cardNumber).toEqual('2345 6789 0876');
	// });

	// it('formats cardExpirationDate state properly when formatCardDetail', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.cardExpirationDate).toEqual('');
	// 	tree.instance().formatCardDetail('cardExpirationDate', '1299');
	// 	expect(tree.instance().state.cardExpirationDate).toEqual('12/99');
	// });

	// it('formats zip state properly when formatCardDetail', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.zip).toEqual('');
	// 	tree.instance().formatCardDetail('zip', '67899');
	// 	expect(tree.instance().state.zip).toEqual('67899');
	// });


	// it('formats cvv state properly when formatCardDetail', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);

	// 	expect(tree.instance().state.cvv).toEqual('');
	// 	tree.instance().formatCardDetail('cvv', '129');
	// 	expect(tree.instance().state.cvv).toEqual('129');
	// });

	// it('return truthy value if form is valid when isFormValid method is called', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);
	// 	tree.instance().setState({
	// 		cardNumber: '23456789',
	// 		cardExpirationDate: '12/09',
	// 		cvv: '437',
	// 		zip: '23457',
	// 	});
	// 	expect(tree.instance().isFormValid()).toEqual(true);
	// });

	// it('return falsy value if form is invalid when isFormValid method is called', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);
	// 	expect(tree.instance().isFormValid()).toEqual(false);
	// });

	// it('should call isFormValid when component did update', () => {
	// 	const tree = shallow(
	// 		<AddPaymentMethodForm
	// 			{...initialProps}
	// 		/>
	// 	);
	// 	tree.instance().isFormValid = jest.fn();
	// 	tree.instance().componentDidUpdate({ prevProps: { isPaymentCardAdded: false } });
	// 	expect(tree.instance().isFormValid).toHaveBeenCalled();
	// });
});
