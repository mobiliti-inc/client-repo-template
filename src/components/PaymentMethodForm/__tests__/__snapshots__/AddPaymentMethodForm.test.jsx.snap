// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPaymentMethodForm renders 1`] = `
<div
  className="form-container"
>
  <TextInput
    changePasswordVisibilityStatus={[Function]}
    className=""
    dark={false}
    disabled={false}
    errorMessage=""
    id="cardNumber"
    inputDidChange={false}
    isValid={true}
    label="Card Number"
    name="cardNumber"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Card Number"
    shouldShowVisibility={false}
    type="text"
    value=""
  />
  <div
    className="column-2"
  >
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage=""
      id="cardExpirationDate"
      inputDidChange={false}
      isValid={true}
      label="Expiration"
      name="cardExpirationDate"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Expiration"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage=""
      id="zip"
      inputDidChange={false}
      isValid={true}
      label="Zip"
      name="zip"
      onBlur={null}
      onChange={[Function]}
      placeholder="Zip"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
  </div>
  <div
    className="column-2"
  >
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage=""
      id="cvv"
      inputDidChange={false}
      isValid={true}
      label="CVV"
      name="cvv"
      onBlur={null}
      onChange={[Function]}
      placeholder="CVV"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
  </div>
  <div />
  <div
    className="column-2 btn-container"
  >
    <div
      className="btn-disabled"
    >
      <Button
        bordered={true}
        buttonType="standard"
        className=""
        customStyles=""
        disabled={true}
        icon={null}
        inverted={false}
        onClick={[Function]}
        ripple={false}
        showLoader={false}
        type="standard"
      >
        Save
      </Button>
    </div>
  </div>
</div>
`;

exports[`AddPaymentMethodForm renders without onFormValidCheck function 1`] = `
<div
  className="form-container"
>
  <TextInput
    changePasswordVisibilityStatus={[Function]}
    className=""
    dark={false}
    disabled={false}
    errorMessage=""
    id="cardNumber"
    inputDidChange={false}
    isValid={true}
    label="Card Number"
    name="cardNumber"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Card Number"
    shouldShowVisibility={false}
    type="text"
    value=""
  />
  <div
    className="column-2"
  >
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage=""
      id="cardExpirationDate"
      inputDidChange={false}
      isValid={true}
      label="Expiration"
      name="cardExpirationDate"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Expiration"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage=""
      id="zip"
      inputDidChange={false}
      isValid={true}
      label="Zip"
      name="zip"
      onBlur={null}
      onChange={[Function]}
      placeholder="Zip"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
  </div>
  <div
    className="column-2"
  >
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage=""
      id="cvv"
      inputDidChange={false}
      isValid={true}
      label="CVV"
      name="cvv"
      onBlur={null}
      onChange={[Function]}
      placeholder="CVV"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
  </div>
  <div />
  <div
    className="column-2 btn-container"
  >
    <div
      className="btn-disabled"
    >
      <Button
        bordered={true}
        buttonType="standard"
        className=""
        customStyles=""
        disabled={true}
        icon={null}
        inverted={false}
        onClick={[Function]}
        ripple={false}
        showLoader={false}
        type="standard"
      >
        Save
      </Button>
    </div>
    <Button
      bordered={true}
      buttonType="standard"
      className=""
      customStyles=""
      disabled={false}
      icon={null}
      inverted={false}
      onClick={[MockFunction]}
      ripple={false}
      showLoader={false}
      type="standard"
    >
      Cancel
    </Button>
  </div>
</div>
`;
